liblwgeom/lwgeom_transform.c

```diff
@@ -126,49 +198,268 @@ lwgeom_transform(LWGEOM *geom, projPJ inpj, projPJ outpj)
+
+static int
+proj_crs_is_swapped(const PJ* pj_crs)
+{
+	PJ *pj_cs = proj_crs_get_coordinate_system(NULL, pj_crs);
+	int axis_count = proj_cs_get_axis_count(NULL, pj_cs);
+	if (axis_count > 0)
+	{
+		const char *out_name, *out_abbrev, *out_direction;
+		double out_unit_conv_factor;
+		const char *out_unit_name, *out_unit_auth_name, *out_unit_code;
+		/* Read only first axis, see if it is degrees / north */
+		proj_cs_get_axis_info(NULL, pj_cs, 0,
+			&out_name,
+			&out_abbrev,
+			&out_direction,
+			&out_unit_conv_factor,
+			&out_unit_name,
+			&out_unit_auth_name,
+			&out_unit_code
+			);
+		return (strcasecmp(out_direction, "north") == 0);
+	}
+	return LW_FALSE;
+}
+
+
+int
+ptarray_transform(POINTARRAY* pa, PJ* pj)
+{
+	uint32_t i;
+	POINT4D p;
+	size_t n_converted;
+	size_t n_points = pa->npoints;
+	size_t point_size = ptarray_point_size(pa);
+	int has_z = ptarray_has_z(pa);
+	double *pa_double = (double*)(pa->serialized_pointlist);
+	int input_swapped, output_swapped;
+
+	/* XXXX TODO check that the PJ has decimal degrees as units in input/output */
+
+	PJ* pj_source_crs = proj_get_source_crs(NULL, pj);
+	PJ* pj_target_crs = proj_get_target_crs(NULL, pj);
+
+	if (!(pj_source_crs && pj_source_crs))
+	{
+		lwerror("ptarray_transform: unable to access source and target crs");
 		return LW_FAILURE;
 	}
 
-	if (pj_is_latlong(dstpj)) to_dec(pt);
+	input_swapped = proj_crs_is_swapped(pj_source_crs);
+	output_swapped = proj_crs_is_swapped(pj_target_crs);
+	proj_destroy(pj_source_crs);
+	proj_destroy(pj_target_crs);
+
+	/* Convert to radians if necessary */
+	if (proj_angular_input(pj, PJ_FWD))
+	{
+		for (i = 0; i < pa->npoints; i++)
+		{
+			getPoint4d_p(pa, i, &p);
+			to_rad(&p);
+		}
+	}
+
+	if (input_swapped)
+		ptarray_swap_ordinates(pa, LWORD_X, LWORD_Y);
+
+	/*
+	* size_t proj_trans_generic(PJ *P, PJ_DIRECTION direction,
+	* double *x, size_t sx, size_t nx,
+	* double *y, size_t sy, size_t ny,
+	* double *z, size_t sz, size_t nz,
+	* double *t, size_t st, size_t nt)
+	*/
+
+	n_converted = proj_trans_generic(
+		pj, PJ_FWD,
+		pa_double,     point_size, n_points, /* X */
+		pa_double + 1, point_size, n_points, /* Y */
+		has_z ? pa_double + 2 : NULL,
+		has_z ? point_size : 0,
+		has_z ? n_points : 0, /* Z */
+		NULL, 0, 0 /* M */
+		);
+
+	if (n_converted != n_points)
+	{
+		lwerror("ptarray_transform: converted (%d) != input (%d)",
+			n_converted, n_points);
+		return LW_FAILURE;
+	}
+
+	int pj_errno_val = proj_errno(pj);
+	if (pj_errno_val)
+	{
+		lwerror("transform: %s (%d)",
+			proj_errno_string(pj_errno_val), pj_errno_val);
+		return LW_FAILURE;
+	}
+
+	if (output_swapped)
+		ptarray_swap_ordinates(pa, LWORD_X, LWORD_Y);
+
+	/* Convert radians to degrees if necessary */
+	if (proj_angular_output(pj, PJ_FWD))
+	{
+		for (i = 0; i < pa->npoints; i++)
+		{
+			getPoint4d_p(pa, i, &p);
+			to_dec(&p);
+		}
+	}
+
 	return LW_SUCCESS;
 }
 
-projPJ
-lwproj_from_string(const char *str1)
+#if 0
+int
+point4d_transform(POINT4D *pt, PJ* pj)
```

Comment:
Since you want to keep it, you need to handle errors in `proj_get_source_crs` as it might return NULL and change the behaviour of proj_crs_is_swapped (the default context will be used, instead of what you'd expect).

It also leaks memory.